File: leetcode/Algorithms/39.CombinationSum/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<List<Integer>>> table = new ArrayList<>();
        for (int i = 0; i <= target; i++) table.add(new ArrayList<>());
        table.get(0).add(new ArrayList<>());

        for (int candidate : candidates) {
            for (int i = 0; i <= target - candidate; i++) {
                for (List<Integer> origin : table.get(i)) {
                    ArrayList<Integer> tmp = new ArrayList<>(origin);
                    tmp.add(candidate);
                    table.get(i + candidate).add(tmp);
                }
            }
        }
        return table.get(target);
    }
}

File: leetcode/Algorithms/39.CombinationSum/Solution.py
iclass Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        self.backtrack(candidates, target, 0, [], result)
        return result

    def backtrack(self, candidates, target, start, tmp, result):
        if target == 0:
            result.append(tmp[:])
        elif target > 0:
            for i in range(start, len(candidates)):
                tmp.append(candidates[i])
                self.backtrack(candidates, target - candidates[i], i, tmp, result)
                tmp.pop()

